src/protocol/sort/reshare.rs

```diff
@@ -0,0 +1,146 @@
+    prss: &'a PrssSpace,
+    record_id: RecordId,
+    step: S,
+}
+
+impl<'a, G, S: Step + SpaceIndex> Reshare<'a, G, S> {
+    #[allow(dead_code)]
+    pub fn new(prss: &'a PrssSpace, gateway: &'a G, record_id: RecordId, step: S) -> Self {
+        Self {
+            gateway,
+            prss,
+            record_id,
+            step,
+        }
+    }
+
+    /// Steps
+    /// 1. While calculating for helper, we call PRSS to get randoms for helper who needs to receive reshares (say r0, r1)
+    ///    `to_helper.left` knows r0 (named r1) and `to_helper.right` knows r1 (named r0)
+    /// 2. `to_helper.left` calculates part1 = (input.0 + input.1) - r1
+    ///    `to_helper.right` calculates part2 = (input.0 - r0)
+    /// 3. `to_helper.left` and `to_helper.right` exchange their calculated shares
+    /// 4. Everyone sets their shares  
+    ///    `to_helper.left` = (r0, part1 + part2))
+    ///    `to_helper.right` = (part1 + part2, r1)
+    ///    `to_helper`      = (r1, r0)
+
+    #[allow(dead_code)]
+    pub async fn execute<F, M>(
+        self,
+        input: Replicated<F>,
+        to_helper: Identity,
+    ) -> Result<Replicated<F>, BoxError>
+    where
+        F: Field + 'static,
+        M: Mesh + 'a,
+        G: Gateway<M, S>,
+    {
+        let mut channel = self.gateway.get_channel(self.step);
+
+        let (r0, r1) = self.prss.generate_fields(self.record_id.into());
+
+        let inputs = input.as_tuple();
+        // `to_helper.left` calculates part1 = (input.0 + input.1) - r1 and sends part1 to `to_helper.right`
+        if channel.identity() == to_helper.peer(Direction::Left) {
+            let part1 = inputs.0 + inputs.1 - r1;
+            channel
+                .send(
+                    to_helper.peer(Direction::Right),
+                    self.record_id,
+                    DValue { d: part1 },
+                )
+                .await?;
+
+            // Sleep until `to_helper.left` sends us their part2 value
+
+            let DValue { d: part2 } = channel
+                .receive(channel.identity().peer(Direction::Left), self.record_id)
+                .await?;
+
+            Ok(Replicated::new(r1, part1 + part2))
+        } else if channel.identity() == to_helper.peer(Direction::Right) {
+            //  `to_helper.right` calculates part2 = (input.0 - r0) and sends it to `to_helper.left`
+            let part2 = inputs.0 - r0;
+            channel
+                .send(
+                    to_helper.peer(Direction::Left),
+                    self.record_id,
+                    DValue { d: part2 },
+                )
+                .await?;
+
+            // Sleep until `to_helper.right` sends us their part1 value
+
+            let DValue { d: part1 }: DValue<F> = channel
+                .receive(channel.identity().peer(Direction::Right), self.record_id)
+                .await?;
+
+            Ok(Replicated::new(part1 + part2, r0))
+        } else {
+            Ok(Replicated::new(r1, r0))
```

Comment:
I don't understand why it is (r1, r0) and not the opposite. It does not match what is shown on the picture in the description - it seems everything is just reversed, what is the reason for that?

Reply:
I got messed up while writing this code. (The picture description is correct). Surprisingly validate_and_reconstruct did not catch this as well. Have added an extra check in that function to ensure Replicated secret shares are same across helpers.

Reply:
I think it worked because **everything** was reversed so math was the same

Reply:
Everything was reversed BUT the replicated secret share construct i.e. P1.left = P3.right etc was not being met. Hence added that check. Such things can become very easy to mess up